File: py_solutions_1-10/Euler_2.py
# sum of even fibonacci numbers below 4 million? 

from __future__ import print_function
import timeit


start = timeit.default_timer()
def euler_2():
    a, b = 0, 1
    while b <= 4000000:
        a, b = b, a + b
        if b % 2 == 0:
            yield b
"""
def euler_2(t=0, n=2, ln=1):
    if n >= 4000000: return t
    elif n % 2 == 0: return euler_2(t + n, n + ln, n)
    return euler_2(t, n + ln, n)
"""

#print("Answer: {}".format(euler_2()))
print("Answer: {}".format(sum([i for i in euler_2()])))
stop = timeit.default_timer()
print("Time: {0:9.5f}".format(stop - start))

File: Haskell/haskell_1-10/Euler_2.hs
-- Total of all even fibonacci numbers below 4 million
module Main where
import Text.Printf
import System.Clock


fib :: Int -> Int -> Int -> Int
fib n1 n2 total
    | n1 > 4000000      = total
    | n1 `mod` 2 == 0   = fib n2 (n2 + n1) total
    | otherwise         = fib n2 (n2 + n1) (total + n1)

getTotalTime :: TimeSpec -> Float
getTotalTime t = fromIntegral (sec t) + (fromIntegral(nsec t) / 10^9)

main :: IO ()
main =
  do
    start <- getTime Monotonic
    printf "Answer:  %d\n" (fib 0 1 0)
    stop <- getTime Monotonic
    printf "Time:    %.6f\n" (getTotalTime(stop) - getTotalTime(start))
